--- ../src-base/minecraft/net/minecraft/block/Block.java
+++ ../src-work/minecraft/net/minecraft/block/Block.java
@@ -1,5 +1,6 @@
 package net.minecraft.block;
 
+import carpet.CarpetSettings;
 import com.google.common.collect.Sets;
 import com.google.common.collect.UnmodifiableIterator;
 import java.util.List;
@@ -48,6 +49,10 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import carpet.helpers.CapturedDrops;
+import carpet.helpers.StatHelper;
+import carpet.patches.BlockWool;
+
 public class Block
 {
     private static final ResourceLocation field_176230_a = new ResourceLocation("air");
@@ -74,7 +79,7 @@
     public float field_149765_K;
     protected final BlockStateContainer field_176227_L;
     private IBlockState field_176228_M;
-    private String field_149770_b;
+    public String field_149770_b;
 
     public static int func_149682_b(Block p_149682_0_)
     {
@@ -236,7 +241,7 @@
         return this;
     }
 
-    protected Block func_149713_g(int p_149713_1_)
+    public Block func_149713_g(int p_149713_1_)
     {
         this.field_149786_r = p_149713_1_;
         return this;
@@ -328,7 +333,7 @@
         return this.field_149782_v;
     }
 
-    protected Block func_149675_a(boolean p_149675_1_)
+    public Block func_149675_a(boolean p_149675_1_)
     {
         this.field_149789_z = p_149675_1_;
         return this;
@@ -400,7 +405,10 @@
 
     public void func_180645_a(World p_180645_1_, BlockPos p_180645_2_, IBlockState p_180645_3_, Random p_180645_4_)
     {
+        if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+            carpet.carpetclient.CarpetClientChunkLogger.setReason("Randomtick block: " + p_180645_3_.func_177230_c().func_149732_F());
         this.func_180650_b(p_180645_1_, p_180645_2_, p_180645_3_, p_180645_4_);
+        carpet.carpetclient.CarpetClientChunkLogger.resetReason();
     }
 
     public void func_180650_b(World p_180650_1_, BlockPos p_180650_2_, IBlockState p_180650_3_, Random p_180650_4_)
@@ -454,46 +462,63 @@
         }
     }
 
+    public final void dropBlockAsItem(World worldIn, BlockPos pos, IBlockState state, int fortune, boolean safeMined)
+    {
+        this.dropBlockAsItemWithChance(worldIn, pos, state, 1.0F, fortune, safeMined);
+    }
+
     public final void func_176226_b(World p_176226_1_, BlockPos p_176226_2_, IBlockState p_176226_3_, int p_176226_4_)
     {
-        this.func_180653_a(p_176226_1_, p_176226_2_, p_176226_3_, 1.0F, p_176226_4_);
+        dropBlockAsItem(p_176226_1_, p_176226_2_, p_176226_3_, p_176226_4_, false);
     }
 
-    public void func_180653_a(World p_180653_1_, BlockPos p_180653_2_, IBlockState p_180653_3_, float p_180653_4_, int p_180653_5_)
+    public void dropBlockAsItemWithChance(World worldIn, BlockPos pos, IBlockState state, float chance, int fortune, boolean safeMined)
     {
-        if (!p_180653_1_.field_72995_K)
+        if (!worldIn.field_72995_K)
         {
-            int i = this.func_149679_a(p_180653_5_, p_180653_1_.field_73012_v);
+            int i = this.func_149679_a(fortune, worldIn.field_73012_v);
 
             for (int j = 0; j < i; ++j)
             {
-                if (p_180653_1_.field_73012_v.nextFloat() <= p_180653_4_)
+                if (worldIn.field_73012_v.nextFloat() <= chance)
                 {
-                    Item item = this.func_180660_a(p_180653_3_, p_180653_1_.field_73012_v, p_180653_5_);
+                    Item item = this.func_180660_a(state, worldIn.field_73012_v, fortune);
 
                     if (item != Items.field_190931_a)
                     {
-                        func_180635_a(p_180653_1_, p_180653_2_, new ItemStack(item, 1, this.func_180651_a(p_180653_3_)));
+                        spawnAsEntity(worldIn, pos, new ItemStack(item, 1, this.func_180651_a(state)), safeMined);
                     }
                 }
             }
         }
     }
 
-    public static void func_180635_a(World p_180635_0_, BlockPos p_180635_1_, ItemStack p_180635_2_)
+    public void func_180653_a(World p_180653_1_, BlockPos p_180653_2_, IBlockState p_180653_3_, float p_180653_4_, int p_180653_5_)
     {
-        if (!p_180635_0_.field_72995_K && !p_180635_2_.func_190926_b() && p_180635_0_.func_82736_K().func_82766_b("doTileDrops"))
-        {
+        dropBlockAsItemWithChance(p_180653_1_, p_180653_2_, p_180653_3_, p_180653_4_, p_180653_5_, false);
+    }
+
+    public static void spawnAsEntity(World worldIn, BlockPos pos, ItemStack stack, boolean safeMined)
+    {
+        if (!worldIn.field_72995_K && !stack.func_190926_b() && worldIn.func_82736_K().func_82766_b("doTileDrops")) {
             float f = 0.5F;
-            double d0 = (double)(p_180635_0_.field_73012_v.nextFloat() * 0.5F) + 0.25D;
-            double d1 = (double)(p_180635_0_.field_73012_v.nextFloat() * 0.5F) + 0.25D;
-            double d2 = (double)(p_180635_0_.field_73012_v.nextFloat() * 0.5F) + 0.25D;
-            EntityItem entityitem = new EntityItem(p_180635_0_, (double)p_180635_1_.func_177958_n() + d0, (double)p_180635_1_.func_177956_o() + d1, (double)p_180635_1_.func_177952_p() + d2, p_180635_2_);
-            entityitem.func_174869_p();
-            p_180635_0_.func_72838_d(entityitem);
+            double d0 = (double) (worldIn.field_73012_v.nextFloat() * 0.5F) + 0.25D;
+            double d1 = (double) (worldIn.field_73012_v.nextFloat() * 0.5F) + 0.25D;
+            double d2 = (double) (worldIn.field_73012_v.nextFloat() * 0.5F) + 0.25D;
+            EntityItem entityitem = new EntityItem(worldIn, (double) pos.func_177958_n() + d0, (double) pos.func_177956_o() + d1, (double) pos.func_177952_p() + d2, stack);
+            if (safeMined) entityitem.func_174868_q();
+            else entityitem.func_174869_p();
+            worldIn.func_72838_d(entityitem);
+            if (CapturedDrops.isCapturingDrops())
+                CapturedDrops.captureDrop(entityitem);
         }
     }
 
+    public static void func_180635_a(World p_180635_0_, BlockPos p_180635_1_, ItemStack p_180635_2_)
+    {
+        spawnAsEntity(p_180635_0_, p_180635_1_, p_180635_2_, false);
+    }
+
     protected void func_180637_b(World p_180637_1_, BlockPos p_180637_2_, int p_180637_3_)
     {
         if (!p_180637_1_.field_72995_K && p_180637_1_.func_82736_K().func_82766_b("doTileDrops"))
@@ -594,18 +619,24 @@
 
     public void func_180657_a(World p_180657_1_, EntityPlayer p_180657_2_, BlockPos p_180657_3_, IBlockState p_180657_4_, @Nullable TileEntity p_180657_5_, ItemStack p_180657_6_)
     {
-        p_180657_2_.func_71029_a(StatList.func_188055_a(this));
+        // CM
+        // player.addStat(StatList.getBlockStats(this));
+        p_180657_2_.func_71029_a(StatHelper.getBlockStateStats(p_180657_4_));
+        // CM END
         p_180657_2_.func_71020_j(0.005F);
 
+        boolean safeMined = CarpetSettings.safeMining && p_180657_2_.func_70093_af();
+        if (safeMined) p_180657_3_ = p_180657_2_.func_180425_c();
+
         if (this.func_149700_E() && EnchantmentHelper.func_77506_a(Enchantments.field_185306_r, p_180657_6_) > 0)
         {
             ItemStack itemstack = this.func_180643_i(p_180657_4_);
-            func_180635_a(p_180657_1_, p_180657_3_, itemstack);
+            spawnAsEntity(p_180657_1_, p_180657_3_, itemstack, safeMined);
         }
         else
         {
             int i = EnchantmentHelper.func_77506_a(Enchantments.field_185308_t, p_180657_6_);
-            this.func_176226_b(p_180657_1_, p_180657_3_, p_180657_4_, i);
+            this.dropBlockAsItem(p_180657_1_, p_180657_3_, p_180657_4_, i, safeMined);
         }
     }
 
@@ -849,7 +880,8 @@
         func_176219_a(32, "deadbush", (new BlockDeadBush()).func_149711_c(0.0F).func_149672_a(SoundType.field_185850_c).func_149663_c("deadbush"));
         func_176219_a(33, "piston", (new BlockPistonBase(false)).func_149663_c("pistonBase"));
         func_176219_a(34, "piston_head", (new BlockPistonExtension()).func_149663_c("pistonBase"));
-        func_176219_a(35, "wool", (new BlockColored(Material.field_151580_n)).func_149711_c(0.8F).func_149672_a(SoundType.field_185854_g).func_149663_c("cloth"));
+        // CM: use custom wool block
+        func_176219_a(35, "wool", (new BlockWool()).func_149711_c(0.8F).func_149672_a(SoundType.field_185854_g).func_149663_c("cloth"));
         func_176219_a(36, "piston_extension", new BlockPistonMoving());
         func_176219_a(37, "yellow_flower", (new BlockYellowFlower()).func_149711_c(0.0F).func_149672_a(SoundType.field_185850_c).func_149663_c("flower1"));
         func_176219_a(38, "red_flower", (new BlockRedFlower()).func_149711_c(0.0F).func_149672_a(SoundType.field_185850_c).func_149663_c("flower2"));
